src/pages/bot-firewall/index.tsx

```diff
@@ -0,0 +1,168 @@
+import { BlockIpPayload, BlockIpResponse } from '../api/bot-firewall/block-ip';
+import { useVisitorData } from '@fingerprintjs/fingerprintjs-pro-react';
+import classnames from 'classnames';
+import { OptionsObject as SnackbarOptions, enqueueSnackbar } from 'notistack';
+import { BOT_FIREWALL_COPY } from '../../client/bot-firewall/botFirewallCopy';
+
+const formatDate = (date: string) => {
+  const d = new Date(date);
+  return `${d.toLocaleDateString('en-Us', {
+    month: 'short',
+    day: 'numeric',
+  })} ${d.toLocaleTimeString('gb', { hour: '2-digit', minute: '2-digit' })}`;
+};
+
+const snackbarOptions: SnackbarOptions = {
+  autoHideDuration: 3000,
+  anchorOrigin: { horizontal: 'right', vertical: 'bottom' },
+};
+
+export const BotFirewall: NextPage<CustomPageProps> = ({ embed }) => {
+  // Get visitor data from Fingerprint (just used for the visitor's IP address)
+  const {
+    getData: getVisitorData,
+    data: visitorData,
+    isLoading: isLoadingVisitorData,
+  } = useVisitorData({
+    extendedResult: true,
+  });
+
+  // Get a list of bot visits
+  const {
+    data: botVisits,
+    refetch: refetchBotVisits,
+    isLoading: isLoadingBotVisits,
+  } = useQuery({
+    queryKey: ['get bot visits'],
+    queryFn: (): Promise<BotVisit[]> => {
+      return fetch('/api/bot-firewall/get-bot-visits').then((res) => res.json());
+    },
+  });
+
+  // Get a list of currently blocked IP addresses
+  const { data: blockedIps, refetch: refetchBlockedIps } = useQuery({
+    queryKey: ['get blocked IPs'],
+    queryFn: (): Promise<string[]> => {
+      return fetch('/api/bot-firewall/get-blocked-ips').then((res) => res.json());
+    },
+  });
+
+  // Post request mutation to block/unblock IP addresses
+  const { mutate: blockIp, isLoading: isLoadingBlockIp } = useMutation({
+    mutationKey: ['block IP'],
+    mutationFn: async ({ ip, blocked }: Omit<BlockIpPayload, 'requestId'>) => {
+      const { requestId } = await getVisitorData({ ignoreCache: true });
+      const response = await fetch('/api/bot-firewall/block-ip', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({ ip, blocked, requestId } satisfies BlockIpPayload),
+      });
+      if (response.ok) {
+        return await response.json();
+      } else {
+        throw new Error('Failed to update firewall: ' + (await response.json()).message ?? response.statusText);
+      }
+    },
+    onSuccess: async (data: BlockIpResponse) => {
+      refetchBlockedIps();
+      enqueueSnackbar(
+        <>
+          IP address <b>&nbsp;{data.ip}&nbsp;</b> was <b>&nbsp;{data.blocked ? 'blocked' : 'unblocked'}&nbsp;</b> in the
+          application firewall.{' '}
+        </>,
+        { ...snackbarOptions, variant: 'success' },
+      );
+    },
+    onError: (error: Error) => {
+      enqueueSnackbar(error.message, {
+        ...snackbarOptions,
+        variant: 'error',
+      });
+    },
+  });
```

Comment:
Could we extract these API calls to separate hooks?

Reply:
Like this? https://github.com/fingerprintjs/fingerprintjs-pro-use-cases/pull/111/commits/9556ad31c031aa51ec70403b2e132a9a93bd503e

Reply:
Yep, looks good!